contracts/style.md

```diff
@@ -0,0 +1,215 @@
+
+## Vendor Dependencies
+
+- That’s it, vendor your Solidity dependencies. Supply chain attacks are all the rage these days. There is not yet a silver bullet for best way to vendor, it depends on the size of your project and your needs. You should be as explicit as possible about where the code comes from and make sure that this is enforced in some way; e.g. reference a hash. Some options:
+    - NPM packages work for repos already in the JavaScript ecosystem. If you go this route you should lock to a hash of the repo or use a proxy registry like GitHub Packages.
+    - Git submodules are great if you don’t mind git submodules.
+    - Copy and paste the code into a `vendor` directory. Record attribution of the code and license in the repo along with the commit or version that you pulled the code from.
+
+## Common Behaviors
+
+### Transferring Ownership
+
+- When transferring control, whether it is of a token or a role in a contract, prefer "safe ownership" transfer patterns where the recipient must accept ownership. This avoids accidentally burning the control. This is also inline with the secure pattern of [prefer pull over push](https://consensys.github.io/smart-contract-best-practices/recommendations/#favor-pull-over-push-for-external-calls).
+
+### Use Factories
+
+- If you expect multiple instances of a contract to be deployed, it is probably best to [build a factory](https://www.quicknode.com/guides/solidity/how-to-create-a-smart-contract-factory-in-solidity-using-hardhat) as this allows for simpler deployments later. Additionally it reduces the burden of verifying the correctness of the contract deployment. If many people have to deploy an instance of a contract then doing so with a contract makes it much easier for verification because instead of checking the code hash and/or the compiler and maybe the source code, you only have to check that the contract was deployed through a factory.
+- Factories can add some pain when deploying with immutable variables. In general it is difficult to parse those out immutable variables from internal transactions. There is nothing inherently wrong with contracts deployed in this manner  but at the time of writing they may not easily verify on Etherscan.
+
+### Call with Exact Gas
+
+- `call` accepts a gas parameter, but that parameter is a ceiling on gas usage. If a transaction does not have enough gas, `call` will simply provide as much gas as it safely can. This is unintuitive and can lead to transactions failing for unexpected reasons. We have [an implementation of `callWithExactGas`](https://github.com/smartcontractkit/chainlink/blob/075f3e2caf61b8685d2dc78714f1ee39764fda17/contracts/src/v0.8/KeeperRegistry.sol#L792) to ensure the precise gas amount requested is provided.
```

Comment:
We had recent discussion in the Solidity guild about this: how `callWithExactGas` should handle return data from the external call. 3 options, depending on use-case:

- Return success only `returns (bool)`
- Return success and return data - without reverting (if `!success`, the return data is the revert data from the external call)
- Return success and revert (bubble the external revert data)

Some implementation examples of these options in this [AddressPlus](https://github.com/alexroan/solidity-library/blob/master/src/AddressPlus.sol) contract.

Reply:
I'm open to changing this, but we should get the version we recommend audited.